File: lc_js/lcJs/design/designLinkedList.ts
// LeetCode 707: 设计链表
// TypeScript版本实现，展示静态类型系统和面向对象特性

// 双向链表节点接口
interface IListNode<T = number> {
    val: T;
    prev: IListNode<T> | null;
    next: IListNode<T> | null;
}

// 链表操作接口
interface ILinkedList<T = number> {
    get(index: number): T | -1;
    addAtHead(val: T): void;
    addAtTail(val: T): void;
    addAtIndex(index: number, val: T): void;
    deleteAtIndex(index: number): void;
    getSize(): number;
}

// 双向链表节点类
class ListNode<T = number> implements IListNode<T> {
    public val: T;
    public prev: ListNode<T> | null = null;
    public next: ListNode<T> | null = null;
    
    constructor(val: T) {
        this.val = val;
    }
    
    toString(): string {
        return `ListNode(${this.val})`;
    }
}

// 双向链表实现
class MyLinkedList<T = number> implements ILinkedList<T> {
    private head: ListNode<T>;
    private tail: ListNode<T>;
    private size: number;
    
    constructor() {
        // 创建哨兵节点
        this.head = new ListNode<T>(null as any);
        this.tail = new ListNode<T>(null as any);
        this.head.next = this.tail;
        this.tail.prev = this.head;
        this.size = 0;
    }
    
    /**
     * 获取链表中第 index 个节点的值
     */
    get(index: number): T | -1 {
        if (index < 0 || index >= this.size) {
            return -1;
        }
        
        let current = this.head.next!;
        for (let i = 0; i < index; i++) {
            current = current.next!;
        }
        return current.val;
    }
    
    /**
     * 在链表第一个元素之前添加一个值为 val 的节点
     */
    addAtHead(val: T): void {
        this.addAtIndex(0, val);
    }
    
    /**
     * 将值为 val 的节点追加到链表的最后一个元素
     */
    addAtTail(val: T): void {
        this.addAtIndex(this.size, val);
    }
    
    /**
     * 在链表中的第 index 个节点之前添加值为 val 的节点
     */
    addAtIndex(index: number, val: T): void {
        if (index < 0 || index > this.size) {
            return;
        }
        
        let pred = this.head;
        for (let i = 0; i < index; i++) {
            pred = pred.next!;
        }
        
        const succ = pred.next!;
        const newNode = new ListNode<T>(val);
        
        newNode.prev = pred;
        newNode.next = succ;
        pred.next = newNode;
        succ.prev = newNode;
        
        this.size++;
    }
    
    /**
     * 如果索引 index 有效，则删除链表中的第 index 个节点
     */
    deleteAtIndex(index: number): void {
        if (index < 0 || index >= this.size) {
            return;
        }
        
        let pred = this.head;
        for (let i = 0; i < index; i++) {
            pred = pred.next!;
        }
        
        const toDelete = pred.next!;
        const succ = toDelete.next!;
        
        pred.next = succ;
        succ.prev = pred;
        
        this.size--;
    }
    
    /**
     * 获取链表大小
     */
    getSize(): number {
        return this.size;
    }
    
    /**
     * 转换为数组
     */
    toArray(): T[] {
        const result: T[] = [];
        let current = this.head.next;
        while (current !== this.tail) {
            result.push(current!.val);
            current = current!.next;
        }
        return result;
    }
    
    /**
     * 字符串表示
     */
    toString(): string {
        return JSON.stringify(this.toArray());
    }
}

// 泛型栈类（用于演示泛型）
class GenericStack<T> {
    private items: T[] = [];
    
    push(item: T): void {
        this.items.push(item);
    }
    
    pop(): T | undefined {
        return this.items.pop();
    }
    
    peek(): T | undefined {
        return this.items[this.items.length - 1];
    }
    
    isEmpty(): boolean {
        return this.items.length === 0;
    }
    
    size(): number {
        return this.items.length;
    }
}

// 使用泛型栈实现的链表
class StackBasedLinkedList<T = number> implements ILinkedList<T> {
    private data: T[] = [];
    
    get(index: number): T | -1 {
        return (index >= 0 && index < this.data.length) ? this.data[index] : -1;
    }
    
    addAtHead(val: T): void {
        this.data.unshift(val);
    }
    
    addAtTail(val: T): void {
        this.data.push(val);
    }
    
    addAtIndex(index: number, val: T): void {
        if (index >= 0 && index <= this.data.length) {
            this.data.splice(index, 0, val);
        }
    }
    
    deleteAtIndex(index: number): void {
        if (index >= 0 && index < this.data.length) {
            this.data.splice(index, 1);
        }
    }
    
    getSize(): number {
        return this.data.length;
    }
    
    toString(): string {
        return JSON.stringify(this.data);
    }
}

// 抽象链表基类
abstract class AbstractLinkedList<T = number> {
    protected abstract size: number;
    
    abstract get(index: number): T | -1;
    abstract addAtHead(val: T): void;
    abstract addAtTail(val: T): void;
    abstract addAtIndex(index: number, val: T): void;
    abstract deleteAtIndex(index: number): void;
    
    getSize(): number {
        return this.size;
    }
    
    isEmpty(): boolean {
        return this.size === 0;
    }
    
    // 模板方法模式
    addMultiple(values: T[]): void {
        values.forEach(val => this.addAtTail(val));
    }
    
    // 抽象方法的默认实现
    clear(): void {
        while (!this.isEmpty()) {
            this.deleteAtIndex(0);
        }
    }
}

// 继承抽象类的具体实现
class ConcreteLinkedList<T = number> extends AbstractLinkedList<T> {
    protected size: number = 0;
    private data: T[] = [];
    
    get(index: number): T | -1 {
        return (index >= 0 && index < this.size) ? this.data[index] : -1;
    }
    
    addAtHead(val: T): void {
        this.data.unshift(val);
        this.size++;
    }
    
    addAtTail(val: T): void {
        this.data.push(val);
        this.size++;
    }
    
    addAtIndex(index: number, val: T): void {
        if (index >= 0 && index <= this.size) {
            this.data.splice(index, 0, val);
            this.size++;
        }
    }
    
    deleteAtIndex(index: number): void {
        if (index >= 0 && index < this.size) {
            this.data.splice(index, 1);
            this.size--;
        }
    }
    
    toString(): string {
        return JSON.stringify(this.data);
    }
}

// 类型别名
type LinkedListOperation<T = number> = 
    | ['get', number]
    | ['addAtHead', T]
    | ['addAtTail', T]
    | ['addAtIndex', number, T]
    | ['deleteAtIndex', number];

type OperationResult<T = number> = T | -1 | null;

// 联合类型和字面量类型
type ListType = 'doubly' | 'singly' | 'array' | 'stack';

// 工厂模式创建链表
class LinkedListFactory {
    static create<T = number>(type: ListType): ILinkedList<T> {
        switch (type) {
            case 'doubly':
                return new MyLinkedList<T>();
            case 'array':
            case 'stack':
                return new StackBasedLinkedList<T>();
            default:
                throw new Error(`Unsupported list type: ${type}`);
        }
    }
}

// 装饰器模式（如果启用了装饰器）
function logOperation(target: any, propertyName: string, descriptor: PropertyDescriptor) {
    const method = descriptor.value;
    
    descriptor.value = function (...args: any[]) {
        console.log(`Calling ${propertyName} with args:`, args);
        const result = method.apply(this, args);
        console.log(`Result:`, result);
        return result;
    };
}

// 使用装饰器的链表类
class LoggedLinkedList<T = number> implements ILinkedList<T> {
    private list: MyLinkedList<T> = new MyLinkedList<T>();
    
    // @logOperation  // 需要启用装饰器支持
    get(index: number): T | -1 {
        return this.list.get(index);
    }
    
    // @logOperation
    addAtHead(val: T): void {
        this.list.addAtHead(val);
    }
    
    // @logOperation
    addAtTail(val: T): void {
        this.list.addAtTail(val);
    }
    
    // @logOperation
    addAtIndex(index: number, val: T): void {
        this.list.addAtIndex(index, val);
    }
    
    // @logOperation
    deleteAtIndex(index: number): void {
        this.list.deleteAtIndex(index);
    }
    
    getSize(): number {
        return this.list.getSize();
    }
}

// 函数式编程风格
const createLinkedListOperations = <T = number>() => {
    const list = new MyLinkedList<T>();
    
    const operations = {
        get: (index: number): T | -1 => list.get(index),
        addAtHead: (val: T): void => { list.addAtHead(val); },
        addAtTail: (val: T): void => { list.addAtTail(val); },
        addAtIndex: (index: number, val: T): void => { list.addAtIndex(index, val); },
        deleteAtIndex: (index: number): void => { list.deleteAtIndex(index); },
        getSize: (): number => list.getSize(),
        toArray: (): T[] => list.toArray()
    } as const;
    
    return operations;
};

// 高阶函数类型
type OperationHandler<T> = (list: ILinkedList<T>) => void;

// 链表操作执行器
class LinkedListExecutor<T = number> {
    private list: ILinkedList<T>;
    
    constructor(list: ILinkedList<T>) {
        this.list = list;
    }
    
    execute(operations: LinkedListOperation<T>[]): OperationResult<T>[] {
        return operations.map(([op, ...params]) => {
            switch (op) {
                case 'get':
                    return this.list.get(params[0] as number);
                case 'addAtHead':
                    this.list.addAtHead(params[0] as T);
                    return null;
                case 'addAtTail':
                    this.list.addAtTail(params[0] as T);
                    return null;
                case 'addAtIndex':
                    this.list.addAtIndex(params[0] as number, params[1] as T);
                    return null;
                case 'deleteAtIndex':
                    this.list.deleteAtIndex(params[0] as number);
                    return null;
                default:
                    return null;
            }
        });
    }
    
    batch(handlers: OperationHandler<T>[]): void {
        handlers.forEach(handler => handler(this.list));
    }
}

// 类型守卫
function isValidIndex(index: unknown): index is number {
    return typeof index === 'number' && index >= 0 && Number.isInteger(index);
}

function isValidValue<T>(value: unknown): value is T {
    return value !== null && value !== undefined;
}

// 安全的链表操作
class SafeLinkedList<T = number> implements ILinkedList<T> {
    private list: MyLinkedList<T> = new MyLinkedList<T>();
    
    get(index: number): T | -1 {
        if (!isValidIndex(index)) {
            throw new TypeError('Index must be a non-negative integer');
        }
        return this.list.get(index);
    }
    
    addAtHead(val: T): void {
        if (!isValidValue(val)) {
            throw new TypeError('Value cannot be null or undefined');
        }
        this.list.addAtHead(val);
    }
    
    addAtTail(val: T): void {
        if (!isValidValue(val)) {
            throw new TypeError('Value cannot be null or undefined');
        }
        this.list.addAtTail(val);
    }
    
    addAtIndex(index: number, val: T): void {
        if (!isValidIndex(index)) {
            throw new TypeError('Index must be a non-negative integer');
        }
        if (!isValidValue(val)) {
            throw new TypeError('Value cannot be null or undefined');
        }
        this.list.addAtIndex(index, val);
    }
    
    deleteAtIndex(index: number): void {
        if (!isValidIndex(index)) {
            throw new TypeError('Index must be a non-negative integer');
        }
        this.list.deleteAtIndex(index);
    }
    
    getSize(): number {
        return this.list.getSize();
    }
}

// 测试接口
interface TestCase<T = number> {
    name: string;
    operations: LinkedListOperation<T>[];
    expected: OperationResult<T>[];
}

// 测试函数
function testDesignLinkedList(): void {
    console.log('=== TypeScript设计链表测试 ===');
    
    // 测试基本双向链表
    console.log('\n--- 基本双向链表测试 ---');
    const list = new MyLinkedList<number>();
    
    list.addAtHead(7);
    list.addAtHead(2);
    list.addAtHead(1);
    console.log(`添加头部元素后: ${list}`);
    
    list.addAtIndex(3, 0);
    console.log(`在索引3添加0后: ${list}`);
    
    list.deleteAtIndex(2);
    console.log(`删除索引2后: ${list}`);
    
    list.addAtHead(6);
    console.log(`添加头部6后: ${list}`);
    
    list.addAtTail(4);
    console.log(`添加尾部4后: ${list}`);
    
    console.log(`获取索引5: ${list.get(5)}`); // -1
    console.log(`获取索引3: ${list.get(3)}`); // 0
    console.log(`获取索引2: ${list.get(2)}`); // 2
    console.log(`获取索引0: ${list.get(0)}`); // 6
    
    // 测试泛型链表
    console.log('\n--- 泛型链表测试 ---');
    const stringList = new MyLinkedList<string>();
    stringList.addAtHead('world');
    stringList.addAtHead('hello');
    stringList.addAtIndex(1, ' ');
    console.log(`字符串链表: ${stringList}`);
    console.log(`获取索引1: ${stringList.get(1)}`); // ' '
    
    // 测试栈基础链表
    console.log('\n--- 栈基础链表测试 ---');
    const stackList = new StackBasedLinkedList<number>();
    stackList.addAtHead(1);
    stackList.addAtTail(3);
    stackList.addAtIndex(1, 2);
    console.log(`栈基础链表: ${stackList}`);
    console.log(`获取索引1: ${stackList.get(1)}`); // 2
    
    // 测试抽象类实现
    console.log('\n--- 抽象类实现测试 ---');
    const concreteList = new ConcreteLinkedList<number>();
    concreteList.addMultiple([1, 2, 3]);
    console.log(`具体链表: ${concreteList}`);
    console.log(`是否为空: ${concreteList.isEmpty()}`); // false
    console.log(`大小: ${concreteList.getSize()}`); // 3
    
    // 测试工厂模式
    console.log('\n--- 工厂模式测试 ---');
    const factoryList = LinkedListFactory.create<number>('doubly');
    factoryList.addAtHead(1);
    factoryList.addAtTail(3);
    factoryList.addAtIndex(1, 2);
    console.log('工厂创建的链表: [1, 2, 3]');
    console.log(`获取索引1: ${factoryList.get(1)}`); // 2
    
    // 测试函数式操作
    console.log('\n--- 函数式操作测试 ---');
    const funcOps = createLinkedListOperations<number>();
    funcOps.addAtHead(1);
    funcOps.addAtTail(3);
    funcOps.addAtIndex(1, 2);
    console.log(`函数式操作结果: ${JSON.stringify(funcOps.toArray())}`);
    console.log(`获取索引1: ${funcOps.get(1)}`); // 2
    
    // 测试操作执行器
    console.log('\n--- 操作执行器测试 ---');
    const executor = new LinkedListExecutor(new MyLinkedList<number>());
    const operations: LinkedListOperation<number>[] = [
        ['addAtHead', 7],
        ['addAtHead', 2],
        ['addAtHead', 1],
        ['addAtIndex', 3, 0],
        ['deleteAtIndex', 2],
        ['get', 2]
    ];
    
    const results = executor.execute(operations);
    console.log(`执行器结果: ${JSON.stringify(results)}`);
    
    // 测试安全链表
    console.log('\n--- 安全链表测试 ---');
    const safeList = new SafeLinkedList<number>();
    try {
        safeList.addAtHead(1);
        safeList.addAtTail(3);
        safeList.addAtIndex(1, 2);
        console.log('安全链表操作成功: [1, 2, 3]');
        console.log(`获取索引1: ${safeList.get(1)}`); // 2
        
        // 测试错误处理
        // safeList.get(-1); // 会抛出TypeError
    } catch (error) {
        console.log(`捕获错误: ${error}`);
    }
    
    // 测试用例
    console.log('\n--- 测试用例执行 ---');
    const testCases: TestCase<number>[] = [
        {
            name: '基本操作测试',
            operations: [
                ['addAtHead', 1],
                ['addAtTail', 3],
                ['addAtIndex', 1, 2],
                ['get', 1],
                ['deleteAtIndex', 1],
                ['get', 1]
            ],
            expected: [null, null, null, 2, null, 3]
        }
    ];
    
    testCases.forEach(testCase => {
        const testExecutor = new LinkedListExecutor(new MyLinkedList<number>());
        const actualResults = testExecutor.execute(testCase.operations);
        const passed = JSON.stringify(actualResults) === JSON.stringify(testCase.expected);
        console.log(`${testCase.name}: ${passed ? '通过' : '失败'}`);
        if (!passed) {
            console.log(`  期望: ${JSON.stringify(testCase.expected)}`);
            console.log(`  实际: ${JSON.stringify(actualResults)}`);
        }
    });
    
    console.log('\n=== TypeScript特性说明 ===');
    console.log('1. 静态类型系统: 编译时类型检查');
    console.log('2. 泛型: 类型参数化，代码复用');
    console.log('3. 接口: 契约定义，结构化类型');
    console.log('4. 抽象类: 部分实现的基类');
    console.log('5. 类型别名: 复杂类型的简化');
    console.log('6. 联合类型: 多种类型的组合');
    console.log('7. 类型守卫: 运行时类型检查');
    console.log('8. 装饰器: 元编程和AOP');
    console.log('9. 模块系统: 代码组织和封装');
    console.log('10. 严格空检查: null/undefined安全');
}

// 导出模块
export {
    IListNode,
    ILinkedList,
    ListNode,
    MyLinkedList,
    StackBasedLinkedList,
    AbstractLinkedList,
    ConcreteLinkedList,
    LinkedListFactory,
    LinkedListExecutor,
    SafeLinkedList,
    GenericStack,
    LoggedLinkedList,
    createLinkedListOperations,
    testDesignLinkedList,
    LinkedListOperation,
    OperationResult,
    ListType,
    TestCase
};

// 默认导出
export default MyLinkedList;

// 如果在Node.js环境中直接运行
if (typeof require !== 'undefined' && require.main === module) {
    testDesignLinkedList();
}

/*
TypeScript实现特点:

1. **静态类型系统**:
   - 编译时类型检查
   - 类型推断
   - 严格空检查

2. **泛型编程**:
   - 类型参数化
   - 约束泛型
   - 泛型接口和类

3. **面向对象特性**:
   - 接口定义
   - 抽象类
   - 继承和多态
   - 访问修饰符

4. **高级类型**:
   - 联合类型
   - 交叉类型
   - 条件类型
   - 映射类型

5. **类型安全**:
   - 类型守卫
   - 断言函数
   - 非空断言

6. **装饰器**:
   - 类装饰器
   - 方法装饰器
   - 属性装饰器

7. **模块系统**:
   - ES6模块
   - 命名空间
   - 声明合并

8. **工具类型**:
   - Partial<T>
   - Required<T>
   - Pick<T, K>
   - Omit<T, K>

算法复杂度:
- get: O(n)
- addAtHead: O(1)
- addAtTail: O(1)
- addAtIndex: O(n)
- deleteAtIndex: O(n)
- 空间复杂度: O(n)
*/

File: lc_rust/src/bin/designLinkedList.rs
// LeetCode 707: 设计链表
// 设计链表的实现。您可以选择使用单链表或双链表。

use std::rc::Rc;
use std::cell::RefCell;

// 双向链表节点
#[derive(Debug, Clone)]
struct ListNode {
    val: i32,
    prev: Option<Rc<RefCell<ListNode>>>,
    next: Option<Rc<RefCell<ListNode>>>,
}

impl ListNode {
    fn new(val: i32) -> Self {
        ListNode {
            val,
            prev: None,
            next: None,
        }
    }
}

// 双向链表实现
#[derive(Debug)]
struct MyLinkedList {
    head: Option<Rc<RefCell<ListNode>>>,
    tail: Option<Rc<RefCell<ListNode>>>,
    size: usize,
}

impl MyLinkedList {
    fn new() -> Self {
        // 创建哨兵节点
        let head = Rc::new(RefCell::new(ListNode::new(0)));
        let tail = Rc::new(RefCell::new(ListNode::new(0)));
        
        head.borrow_mut().next = Some(tail.clone());
        tail.borrow_mut().prev = Some(head.clone());
        
        MyLinkedList {
            head: Some(head),
            tail: Some(tail),
            size: 0,
        }
    }
    
    fn get(&self, index: i32) -> i32 {
        if index < 0 || index as usize >= self.size {
            return -1;
        }
        
        let mut current = self.head.as_ref().unwrap().borrow().next.clone();
        for _ in 0..index {
            if let Some(node) = current {
                current = node.borrow().next.clone();
            }
        }
        
        current.map_or(-1, |node| node.borrow().val)
    }
    
    fn add_at_head(&mut self, val: i32) {
        self.add_at_index(0, val);
    }
    
    fn add_at_tail(&mut self, val: i32) {
        self.add_at_index(self.size as i32, val);
    }
    
    fn add_at_index(&mut self, index: i32, val: i32) {
        if index < 0 || index as usize > self.size {
            return;
        }
        
        let mut pred = self.head.clone();
        for _ in 0..index {
            if let Some(node) = pred {
                pred = node.borrow().next.clone();
            }
        }
        
        if let Some(pred_node) = pred {
            let succ = pred_node.borrow().next.clone();
            let new_node = Rc::new(RefCell::new(ListNode::new(val)));
            
            new_node.borrow_mut().prev = Some(pred_node.clone());
            new_node.borrow_mut().next = succ.clone();
            
            pred_node.borrow_mut().next = Some(new_node.clone());
            if let Some(succ_node) = succ {
                succ_node.borrow_mut().prev = Some(new_node);
            }
            
            self.size += 1;
        }
    }
    
    fn delete_at_index(&mut self, index: i32) {
        if index < 0 || index as usize >= self.size {
            return;
        }
        
        let mut pred = self.head.clone();
        for _ in 0..index {
            if let Some(node) = pred {
                pred = node.borrow().next.clone();
            }
        }
        
        if let Some(pred_node) = pred {
            // 这里立刻拿到 pred_node.next，Ref 在这行就会 drop
            let to_delete = pred_node.borrow().next.clone();
        
            if let Some(to_delete) = to_delete {
                let succ = to_delete.borrow().next.clone();
        
                // 现在 Ref 已经 drop，可以安全 borrow_mut
                pred_node.borrow_mut().next = succ.clone();
                if let Some(succ_node) = succ {
                    succ_node.borrow_mut().prev = Some(pred_node.clone());
                }
        
                self.size -= 1;
            }
        }
        
    }
}

// 简化版本：使用Vec实现
#[derive(Debug)]
struct SimpleLinkedList {
    data: Vec<i32>,
}

impl SimpleLinkedList {
    fn new() -> Self {
        SimpleLinkedList {
            data: Vec::new(),
        }
    }
    
    fn get(&self, index: i32) -> i32 {
        if index < 0 || index as usize >= self.data.len() {
            -1
        } else {
            self.data[index as usize]
        }
    }
    
    fn add_at_head(&mut self, val: i32) {
        self.data.insert(0, val);
    }
    
    fn add_at_tail(&mut self, val: i32) {
        self.data.push(val);
    }
    
    fn add_at_index(&mut self, index: i32, val: i32) {
        if index < 0 || index as usize > self.data.len() {
            return;
        }
        self.data.insert(index as usize, val);
    }
    
    fn delete_at_index(&mut self, index: i32) {
        if index >= 0 && (index as usize) < self.data.len() {
            self.data.remove(index as usize);
        }
    }
}

// 函数式编程风格的链表操作
mod functional {
    use super::*;
    
    pub fn create_list() -> MyLinkedList {
        MyLinkedList::new()
    }
    
    pub fn list_operations(operations: Vec<(&str, Vec<i32>)>) -> Vec<Option<i32>> {
        let mut list = MyLinkedList::new();
        let mut results = Vec::new();
        
        for (op, params) in operations {
            match op {
                "get" => {
                    if let Some(&index) = params.first() {
                        results.push(Some(list.get(index)));
                    }
                }
                "addAtHead" => {
                    if let Some(&val) = params.first() {
                        list.add_at_head(val);
                        results.push(None);
                    }
                }
                "addAtTail" => {
                    if let Some(&val) = params.first() {
                        list.add_at_tail(val);
                        results.push(None);
                    }
                }
                "addAtIndex" => {
                    if params.len() >= 2 {
                        list.add_at_index(params[0], params[1]);
                        results.push(None);
                    }
                }
                "deleteAtIndex" => {
                    if let Some(&index) = params.first() {
                        list.delete_at_index(index);
                        results.push(None);
                    }
                }
                _ => results.push(None),
            }
        }
        
        results
    }
}

fn test_design_linked_list() {
    println!("=== 设计链表测试 ===");
    
    // 测试双向链表实现
    let mut list = MyLinkedList::new();
    
    println!("\n--- 双向链表实现测试 ---");
    list.add_at_head(7);
    list.add_at_head(2);
    list.add_at_head(1);
    println!("添加头部元素后: [1, 2, 7]");
    
    list.add_at_index(3, 0);
    println!("在索引3添加0后: [1, 2, 7, 0]");
    
    list.delete_at_index(2);
    println!("删除索引2后: [1, 2, 0]");
    
    list.add_at_head(6);
    println!("添加头部6后: [6, 1, 2, 0]");
    
    list.add_at_tail(4);
    println!("添加尾部4后: [6, 1, 2, 0, 4]");
    
    println!("获取索引5: {}", list.get(5)); // -1
    println!("获取索引3: {}", list.get(3)); // 0
    println!("获取索引2: {}", list.get(2)); // 2
    println!("获取索引0: {}", list.get(0)); // 6
    
    list.delete_at_index(0);
    println!("删除索引0后，获取索引0: {}", list.get(0)); // 1
    
    // 测试简化版本
    println!("\n--- 简化版本测试 ---");
    let mut simple_list = SimpleLinkedList::new();
    simple_list.add_at_head(1);
    simple_list.add_at_tail(3);
    simple_list.add_at_index(1, 2);
    println!("简化版本操作后: [1, 2, 3]");
    println!("获取索引1: {}", simple_list.get(1)); // 2
    
    simple_list.delete_at_index(1);
    println!("删除索引1后，获取索引1: {}", simple_list.get(1)); // 3
    
    // 测试函数式编程风格
    println!("\n--- 函数式编程风格测试 ---");
    let operations = vec![
        ("addAtHead", vec![7]),
        ("addAtHead", vec![2]),
        ("addAtHead", vec![1]),
        ("addAtIndex", vec![3, 0]),
        ("deleteAtIndex", vec![2]),
        ("addAtHead", vec![6]),
        ("addAtTail", vec![4]),
        ("get", vec![4]),
        ("addAtHead", vec![4]),
        ("addAtIndex", vec![5, 0]),
        ("addAtHead", vec![6]),
    ];
    
    let results = functional::list_operations(operations);
    println!("函数式操作结果: {:?}", results);
    
    println!("\n=== Rust特性说明 ===");
    println!("1. 内存安全: 编译时防止悬垂指针和内存泄漏");
    println!("2. 所有权系统: 自动管理内存，无需手动释放");
    println!("3. Rc<RefCell<>>: 允许多重引用和内部可变性");
    println!("4. 模式匹配: 安全处理Option类型");
    println!("5. 零成本抽象: 高级特性不影响运行时性能");
}

fn main() {
    test_design_linked_list();
}

/*
Rust实现特点:

1. **内存安全**:
   - 编译时检查防止悬垂指针
   - 自动内存管理
   - 无数据竞争

2. **所有权系统**:
   - 每个值都有唯一所有者
   - 借用检查器确保内存安全
   - RAII自动资源管理

3. **智能指针**:
   - Rc<RefCell<>>允许多重引用
   - RefCell提供内部可变性
   - 运行时借用检查

4. **模式匹配**:
   - Option类型安全处理空值
   - match表达式穷尽性检查
   - if let简化模式匹配

5. **零成本抽象**:
   - 高级特性编译时优化
   - 无运行时开销
   - 性能可预测

算法复杂度:
- get: O(n)
- addAtHead: O(1)
- addAtTail: O(1)
- addAtIndex: O(n)
- deleteAtIndex: O(n)
- 空间复杂度: O(n)
*/